rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,777 @@
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingPublish);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::SUBACK(ref suback) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(suback.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingSubscribe);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::UNSUBACK(ref unsuback) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(unsuback.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingUnsubscribe);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::UNASSIGNED => {
+                let mut tx = self.new_tx(msg, toclient);
+                tx.complete = true;
+                &mut MQTTState::set_event(&mut tx, MQTTEvent::UnassignedMsgtype);
+                self.transactions.push(tx);
+            },
+            _ => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                let mut tx = self.new_tx(msg, toclient);
+                tx.complete = true;
+                self.transactions.push(tx);
+            }
+        }
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> bool {
+        if input.len() == 0 {
+            return true;
+        }
+
+        self.request_buffer.extend(input);
```

Comment:
Why do we need to bufferize ?
Can we not use the `AppLayerResult::Incomplete` ?

Reply:
True, this seems to be redundant now.